File: DASTRUCT/Java/InfixPostfix/src/Stack.java
class Node {
    Node previous;
    char data;
    Node next;

    public Node(char data) {
        this.data = data;
    }
}

public class Stack {
    private Node tail;
    private Node head;

    public boolean push(char c) {
        Node newNode = new Node(c);
        newNode.previous = null;
        newNode.next = null;

        if (head == null) {
            head = tail = newNode;
        } else {
            newNode.previous = tail;
            tail.next = newNode;
            tail = tail.next;
        }
        return true;
    }

    public char pop() {
        char data;
        if (isEmpty()) {
            return '\0';
        }
        data = peek();
        if (head == tail) {
            head = tail = null;
        } else {
            tail = tail.previous;
        }
        if (tail != null) { // Null check to prevent NullPointerExceptions
            tail.next = null;
        }
        return data;
    }

    public char peek() {
        if (isEmpty()) {
            return '\0'; // Returns null byte character to prevent reading from null
        }
        return tail.data;
    }

    public boolean isEmpty() {
        return tail == null;
    }

    public void displayAll() {
        Node curr = head;

        while (curr != null) {
            System.out.print(curr.data);
            curr = curr.next;
        }
        System.out.println();
    }
}

File: DASTRUCT/TypeScript/stack_ts/src/Stack.ts
class MyNode<T> {
    value: T;
    next: MyNode<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
    }
}

class MyStack<T> {
    head: MyNode<T> | null;
    tail: MyNode<T> | null;
    size: number;

    constructor();
    constructor(node: MyNode<T>);
    constructor(node?: MyNode<T>) {
        if (node) {
            this.head = node;
            this.tail = node;
            this.size = 1;
        } else {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }
    }

    isEmpty(): boolean {
        return this.head === null;
    }

    push(nodeOrValue: MyNode<T> | T): boolean {
        const node: MyNode<T> = nodeOrValue instanceof MyNode ? nodeOrValue
            : new MyNode(nodeOrValue);

        if (this.isEmpty()) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head = node;
        }

        this.size += 1;

        return true;
    }

    pop(): MyNode<T> | null {
        if (this.isEmpty()) {
            return null;
        }

        const poppedNode: MyNode<T> | null = this.head;
        this.head = this.head!.next;

        if (this.head === null) {
            this.tail = null;
        }
        this.size -= 1;

        poppedNode!.next = null;
        return poppedNode;
    }

    peek(): MyNode<T> | null {
        return this.head;
    }

    display(): void {
        console.log(this);
    }

    getCount(): number {
        return this.size;
    }

    toString(): string {
        let result: string = '';
        let curr: MyNode<T> | null = this.head;

        while (curr !== null) {
            result += String(curr.value) + (curr.next ? ' -> ' : '');
            curr = curr.next;
        }

        return result;
    }
}

export { MyNode, MyStack };